Wlifetime-safety-suggestions fixit puts attribute into wrong places

One case when this happens is unnamed function parameters:

```shell
$ cat a.h
#include <ostream>
struct A {};
std::ostream& operator<<(std::ostream&, const A&);
$ cat a.cc
#include "a.h"
std::ostream& operator<<(std::ostream& str, const A&) {
  return str;
}
$ clang++ -Wlifetime-safety -Wlifetime-safety-suggestions -Xclang -fixit a.cc
In file included from a.cc:1:
./a.h:3:26: warning: parameter in cross-TU function should be marked [[clang::lifetimebound]] [-Wlifetime-safety-cross-tu-suggestions]
    3 | std::ostream& operator<<(std::ostream&, const A&);
      |                          ^~~~~~~~~~~~~
      |                                        [[clang::lifetimebound]]
./a.h:3:26: note: FIX-IT applied suggested code changes
a.cc:3:10: note: param returned here
    3 |   return str;
      |          ^~~
1 warning generated.
$ cat a.h
#include <ostream>
struct A {};
std::ostream& operator<<(std::ostream& [[clang::lifetimebound]], const A&);
$ clang++ a.cc
In file included from a.cc:1:
./a.h:3:42: error: 'clang::lifetimebound' attribute only applies to parameters and implicit object parameters
    3 | std::ostream& operator<<(std::ostream& [[clang::lifetimebound]], const A&);
      |                                          ^
```

Another case is a method with a trailing return type - the attribute gets added after the return type where it's not recognized, e.g.:

```c++
auto Foo() -> decltype(0) [[clang::lifetimebound]] {  // error: 'clang::lifetimebound' attribute only applies to parameters and implicit object parameters
```

--- a/clang/lib/Sema/AnalysisBasedWarnings.cpp
+++ b/clang/lib/Sema/AnalysisBasedWarnings.cpp
@@ -2937,10 +2937,16 @@ public:
             : diag::warn_lifetime_safety_intra_tu_param_suggestion;
     SourceLocation InsertionPoint = Lexer::getLocForEndOfToken(
         ParmToAnnotate->getEndLoc(), 0, S.getSourceManager(), S.getLangOpts());
+    StringRef FixItText = " [[clang::lifetimebound]]";
+    if (!ParmToAnnotate->getIdentifier()) {
+      // For unnamed parameters, placing attributes after the type would be
+      // parsed as a type attribute, not a parameter attribute.
+      InsertionPoint = ParmToAnnotate->getBeginLoc();
+      FixItText = "[[clang::lifetimebound]] ";
+    }
     S.Diag(ParmToAnnotate->getBeginLoc(), DiagID)
         << ParmToAnnotate->getSourceRange()
-        << FixItHint::CreateInsertion(InsertionPoint,
-                                      " [[clang::lifetimebound]]");
+        << FixItHint::CreateInsertion(InsertionPoint, FixItText);
     S.Diag(EscapeExpr->getBeginLoc(),
            diag::note_lifetime_safety_suggestion_returned_here)
         << EscapeExpr->getSourceRange();
@@ -2952,10 +2958,24 @@ public:
     unsigned DiagID = (Scope == SuggestionScope::CrossTU)
                           ? diag::warn_lifetime_safety_cross_tu_this_suggestion
                           : diag::warn_lifetime_safety_intra_tu_this_suggestion;
-    SourceLocation InsertionPoint;
-    InsertionPoint = Lexer::getLocForEndOfToken(
-        MD->getTypeSourceInfo()->getTypeLoc().getEndLoc(), 0,
-        S.getSourceManager(), S.getLangOpts());
+    const auto MDL = MD->getTypeSourceInfo()->getTypeLoc();
+    SourceLocation InsertionPoint = Lexer::getLocForEndOfToken(
+        MDL.getEndLoc(), 0, S.getSourceManager(), S.getLangOpts());
+    if (const auto *FPT = MD->getType()->getAs<FunctionProtoType>();
+        FPT && FPT->hasTrailingReturn()) {
+      // For trailing return types, 'getEndLoc()' includes the return type
+      // after '->', placing the attribute in an invalid position.
+      // Instead use 'getLocalRangeEnd()' which gives the '->' location
+      // for trailing returns, so find the last token before it.
+      const auto FTL = MDL.getAs<FunctionTypeLoc>();
+      assert(FTL);
+      InsertionPoint = Lexer::getLocForEndOfToken(
+          Lexer::findPreviousToken(FTL.getLocalRangeEnd(), S.getSourceManager(),
+                                   S.getLangOpts(),
+                                   /*IncludeComments=*/false)
+              ->getLocation(),
+          0, S.getSourceManager(), S.getLangOpts());
+    }
     S.Diag(InsertionPoint, DiagID)
         << MD->getNameInfo().getSourceRange()
         << FixItHint::CreateInsertion(InsertionPoint,

--- a/clang/test/Sema/warn-lifetime-safety-fixits.cpp
+++ b/clang/test/Sema/warn-lifetime-safety-fixits.cpp
@@ -138,3 +138,27 @@ const OutOfLine &OutOfLine::get() const {
   // CHECK: fix-it:"{{.*}}":{[[@LINE-2]]:40-[[@LINE-2]]:40}:" {{\[\[}}clang::lifetimebound]]"
   return *this;
 }
+
+struct TrailingReturn {
+  TrailingReturn() {}
+  ~TrailingReturn() {}
+  MyObj data;
+
+  auto get_view() -> View {
+    // CHECK: :[[@LINE-1]]:18: warning: implicit this in intra-TU function should be marked
+    // CHECK: fix-it:"{{.*}}":{[[@LINE-2]]:18-[[@LINE-2]]:18}:" {{\[\[}}clang::lifetimebound]]"
+    return data;
+  }
+
+  auto get_view_const() const -> View {
+    // CHECK: :[[@LINE-1]]:30: warning: implicit this in intra-TU function should be marked
+    // CHECK: fix-it:"{{.*}}":{[[@LINE-2]]:30-[[@LINE-2]]:30}:" {{\[\[}}clang::lifetimebound]]"
+    return data;
+  }
+
+  auto get_ref() const -> const MyObj & {
+    // CHECK: :[[@LINE-1]]:23: warning: implicit this in intra-TU function should be marked
+    // CHECK: fix-it:"{{.*}}":{[[@LINE-2]]:23-[[@LINE-2]]:23}:" {{\[\[}}clang::lifetimebound]]"
+    return data;
+  }
+};

--- a/clang/test/Sema/warn-lifetime-safety-suggestions.cpp
+++ b/clang/test/Sema/warn-lifetime-safety-suggestions.cpp
@@ -50,6 +50,9 @@ inline View redeclared_in_header(View a) {  // expected-warning {{parameter in i
   return a;                                 // expected-note {{param returned here}}
 }
 
+View return_unnamed_view(View);                // expected-warning {{parameter in cross-TU function should be marked [[clang::lifetimebound]]}}
+MyObj& return_unnamed_ref(MyObj&, bool);       // expected-warning {{parameter in cross-TU function should be marked [[clang::lifetimebound]]}}
+
 struct ReturnThis {
   const ReturnThis& get() const;           // expected-warning {{implicit this in cross-TU function should be marked [[clang::lifetimebound]]}}.
 };
@@ -88,6 +91,14 @@ int* return_pointer_directly(int* a) {
 
 MyObj* return_pointer_object(MyObj* a) {
   return a;                                // expected-note {{param returned here}} 
+}
+
+View return_unnamed_view(View a) {
+  return a;                               // expected-note {{param returned here}}
+}
+
+MyObj& return_unnamed_ref(MyObj& a, bool c) {
+  return a;                               // expected-note {{param returned here}}
 }
 
 MyObj& return_reference(MyObj& a, // expected-warning {{parameter in intra-TU function should be marked [[clang::lifetimebound]]}}
@@ -343,6 +354,26 @@ MyObj* test_template_inference_with_stack() {
                                           // expected-note@-1 {{returned here}}                                       
 }
 } // namespace inference_with_templates
+
+namespace trailing_return {
+struct TrailingReturn {
+  TrailingReturn() {}
+  ~TrailingReturn() {}
+  MyObj data;
+
+  auto get_view() -> View {                   // expected-warning {{implicit this in intra-TU function should be marked [[clang::lifetimebound]]}}
+    return data;                              // expected-note {{param returned here}}
+  }
+
+  auto get_view_const() const -> View {       // expected-warning {{implicit this in intra-TU function should be marked [[clang::lifetimebound]]}}
+    return data;                              // expected-note {{param returned here}}
+  }
+
+  auto get_ref() const -> const MyObj& {      // expected-warning {{implicit this in intra-TU function should be marked [[clang::lifetimebound]]}}
+    return data;                              // expected-note {{param returned here}}
+  }
+};
+} // namespace trailing_return
 
 //===----------------------------------------------------------------------===//
 // Negative Test Cases

